File: windows/src/AgentIsland/Usage/AppUsage.cs
namespace AgentIsland.Usage;

/// One rate-limit window as rendered in the UI. UsedPercent is normalized to
/// [0, 1]; Error carries the caption shown in place of a value.
/// PeriodSeconds is the window length as reported by the provider (Codex
/// sends limit_window_seconds; Claude's windows are the documented 5h/7d).
/// It drives display labels, so when a provider changes its quota model —
/// Codex replaced its 5-hour window with a single weekly one in July 2026 —
/// the tile relabels itself instead of lying under a hardcoded "5h". null on
/// old cached snapshots; labels fall back to the slot name.
public sealed record WindowUsage(
    double UsedPercent, DateTimeOffset? ResetAt, string? Error, double? PeriodSeconds = null)
{
    public static WindowUsage Unknown { get; } = new(0, null, "no data");

    public bool HasError => !string.IsNullOrEmpty(Error);

    /// A multi-day window (weekly-style) as opposed to an intra-day one.
    public bool IsLongPeriod => (PeriodSeconds ?? 0) >= 2 * 86400;
}

/// A banked Codex rate-limit reset ("reset card") — OpenAI's own title
/// ("Full reset") plus when it stops being usable.
public sealed record ResetCard(string Id, string Title, DateTimeOffset? ExpiresAt);

/// Usage for one provider: the two window slots plus the plan tier ("max",
/// "pro", ...) surfaced as the chip badge next to the provider name.
/// ResetCards carries Codex's banked rate-limit resets (the count OpenAI
/// reports in rate_limit_reset_credits.available_count); ResetCardDetails
/// the per-card title/expiry from wham/rate-limit-reset-credits. null =
/// provider doesn't report any (Claude, old caches).
public sealed record AppUsage(
    WindowUsage FiveHour,
    WindowUsage Weekly,
    string? Plan = null,
    int? ResetCards = null,
    IReadOnlyList<ResetCard>? ResetCardDetails = null)
{
    public static AppUsage Empty { get; } = new(
        new WindowUsage(0, null, null),
        new WindowUsage(0, null, null));

    public static AppUsage ErrorPair(string message) => new(
        new WindowUsage(0, null, message),
        new WindowUsage(0, null, message));

    /// True when this fetch produced nothing usable — both windows errored
    /// and no percentage survived. Used to keep showing the last good value.
    public bool IsErrorOnly =>
        FiveHour.HasError && Weekly.HasError && FiveHour.UsedPercent == 0 && Weekly.UsedPercent == 0;

    /// True when the provider reported only ONE window on a healthy fetch —
    /// Codex's July 2026 shape (a single weekly quota, secondary_window
    /// gone). The UI hides the empty tile instead of pinning "no data".
    /// Cold start (both unknown) and fetch failures don't qualify, because
    /// the primary slot carries an error too.
    public bool SecondaryMissing => Weekly.Error == "no data" && FiveHour.Error is null;
}

File: Sources/Usage/AppUsage.swift
import Foundation

/// One rate-limit window (e.g. Claude's 5h, Codex's 7d). usedPercent is
/// normalized to 0...1 regardless of what the upstream API returns.
struct WindowUsage: Codable {
    let usedPercent: Double
    let resetAt: Date?
    let error: String?
    /// Window length as reported by the provider (Codex sends
    /// limit_window_seconds; Claude's windows are the documented 5h/7d).
    /// Drives display labels, so when a provider changes its quota model —
    /// Codex replaced its 5-hour window with a single weekly one in July
    /// 2026 — the tile relabels itself instead of lying under a hardcoded
    /// "5h". nil on old cached snapshots; labels fall back to the slot name.
    var periodSeconds: TimeInterval? = nil

    static let unknown = WindowUsage(usedPercent: 0, resetAt: nil, error: "no data")

    var percentInt: Int { Int((usedPercent * 100).rounded()) }

    /// A multi-day window (weekly-style) as opposed to an intra-day one.
    var isLongPeriod: Bool { (periodSeconds ?? 0) >= 2 * 86400 }
}

/// One banked Codex reset: OpenAI's own title ("Full reset") + expiry.
/// From `wham/rate-limit-reset-credits`, available cards only.
struct ResetCard: Codable, Identifiable {
    var id: String
    var title: String
    var expiresAt: Date?
}

struct AppUsage: Codable {
    var fiveHour: WindowUsage
    var weekly: WindowUsage
    /// Provider-reported plan tier — Claude's `subscriptionType` (free/pro/max)
    /// or Codex's `plan_type` (free/plus/pro). nil when unknown.
    var plan: String?
    /// Codex banked rate-limit resets ("reset cards") — the count OpenAI
    /// reports in `rate_limit_reset_credits.available_count`. In the
    /// weekly-only quota era these are the user's escape hatches, so the
    /// panel surfaces the count. nil = provider doesn't report any (Claude,
    /// old caches).
    var resetCards: Int? = nil
    /// Per-card detail (what each card is + when it expires). nil when the
    /// detail fetch failed — the count above still stands on its own.
    var resetCardDetails: [ResetCard]? = nil

    init(fiveHour: WindowUsage, weekly: WindowUsage, plan: String? = nil,
         resetCards: Int? = nil, resetCardDetails: [ResetCard]? = nil) {
        self.fiveHour = fiveHour
        self.weekly = weekly
        self.plan = plan
        self.resetCards = resetCards
        self.resetCardDetails = resetCardDetails
    }

    static let empty = AppUsage(fiveHour: .unknown, weekly: .unknown)

    /// True when the provider reported only ONE window on a healthy fetch —
    /// Codex's July 2026 shape (a single weekly quota, secondary_window
    /// gone). The UI hides the empty tile instead of pinning "no data".
    /// Cold start (both slots unknown) still shows both tiles. But a mere
    /// CAPTION on the primary — "start Antigravity to read quota" — must
    /// not resurrect the placeholder: requiring a clean primary here is
    /// what leaked a ghost "周 0%" tile beside every single-pool guest
    /// whose row carried a status line (owner report, 2026-08-09).
    var secondaryMissing: Bool { weekly.error == "no data" && fiveHour.error != "no data" }

    /// Placeholder values shown when a provider is toggled off. Non-zero
    /// so the chart vocabulary stays visible (a 0% ring reads as broken,
    /// a 45% ring reads as "data we're choosing not to surface").
    static let dummy = AppUsage(
        fiveHour: WindowUsage(usedPercent: 0.45, resetAt: nil, error: nil),
        weekly: WindowUsage(usedPercent: 0.28, resetAt: nil, error: nil)
    )
}
